components/fxa-client/ios/FxAClient/FxAccountManager.swift

```diff
@@ -0,0 +1,584 @@
+                self.state = nextState
+                toProcess = self.stateActions(forState: self.state, via: e)
+                if let successiveEvent = toProcess {
+                    FxALog.debug(
+                        "Ran \(e) side-effects for state \(self.state), got successive event \(successiveEvent)."
+                    )
+                }
+            }
+            completionHandler()
+        }
+    }
+
+    // State transition matrix. Returns nil if there's no transition.
+    internal static func nextState(state: AccountState, event: Event) -> AccountState? {
+        switch state {
+        case .start:
+            switch event {
+            case .initialize: return .start
+            case .accountNotFound: return .notAuthenticated
+            case .accountRestored: return .authenticatedNoProfile
+            default: return nil
+            }
+        case .notAuthenticated:
+            switch event {
+            case .authenticated: return .authenticatedNoProfile
+            default: return nil
+            }
+        case .authenticatedNoProfile:
+            switch event {
+            case .authenticationError: return .authenticationProblem
+            case .fetchProfile: return .authenticatedNoProfile
+            case .fetchedProfile: return .authenticatedWithProfile
+            case .failedToFetchProfile: return .authenticatedNoProfile
+            case .logout: return .notAuthenticated
+            default: return nil
+            }
+        case .authenticatedWithProfile:
+            switch event {
+            case .authenticationError: return .authenticationProblem
+            case .logout: return .notAuthenticated
+            default: return nil
+            }
+        case .authenticationProblem:
+            switch event {
+            case .recoveredFromAuthenticationProblem: return .authenticatedNoProfile
+            case .authenticated: return .authenticatedNoProfile
+            case .logout: return .notAuthenticated
+            default: return nil
+            }
+        }
+    }
+
+    // swiftlint:disable function_body_length
+    internal func stateActions(forState: AccountState, via: Event) -> Event? {
+        switch forState {
+        case .start: do {
+            switch via {
+            case .initialize: do {
+                if let acct = tryRestoreAccount() {
+                    account = acct
+                    return Event.accountRestored
+                } else {
+                    return Event.accountNotFound
+                }
+            }
+            default: return nil
+            }
+        }
+        case .notAuthenticated: do {
+            switch via {
+            case .logout: do {
+                // Clean up internal account state and destroy the current FxA device record.
+                do {
+                    try requireAccount().disconnectSync()
```

Comment:
I don't have any suggestion for what to do about it, but it's unfortunately that this reads like "disconnect sync" in the sense of "disconnecting from Sync-the-feature"...